doc/developer/design/20320328_persist_columnar.md

```diff
@@ -0,0 +1,940 @@
+space reduction in dictionary encoding, but this is offset by the overhead of
+the struct itself.
+
+</td>
+
+<tr>
+<td> Jsonb </td>
+<td>
+
+```rust
+// JsonbRef<'a>(Datum<'a>)
+enum Value {
+  Null,
+  Boolean(bool),
+  Number(f64),
+  String(String),
+  Array(Vec<Value>),
+  Map(BTree<String, Value>),
+}
+```
+
+</td>
+<td>
+
+`BinaryArray`
+
+</td>
+<td>
+
+Serialize JSON with the existing protobuf types, i.e. ProtoDatum, and store
+this binary blob.
+
+> **Structured Data:** An option is to structure the JSON data using an Arrow
+Union type. What is nice about this approach is it would allow us to do some
+form of projection pushdown on the JSON data. The main issue though is Arrow
+does not really support recursive data types. In fact, it is impossible to
+statically define the above `Value` enum in Arrow. The only option is to
+dynamically generate a DataType/schema given a column of values, see [1] for an
+example of this approach. I don't believe dynamically generating the schema is
+a good option because it is relatively complex, and we would end up with
+arbitrarily deep schemas based on user provided data. The arbitrarily deep
+schemas particularly concerns me because it would have unpredictable
+performance.
+
+> **Alternative:** An alternative to fully structing the data is structuing it
+with a depth limit. For example, structuring up-to X levels deep, and then
+binary encoding the rest. This gets us predictable performance with the ability
+to do limited pushdown, at the cost of code complexity. This is probably the
+best approach in the long term, but in my opinion the additional technical
+complexity makes it out-of-scope for the initial implementation.
+
+> **Alternative:** Instead of serializing the JSON data with protobuf, we could
+use a different serialization format like [BSON](https://bsonspec.org/). This
+approach is nice because it gets us a path to entirely eliminating `ProtoDatum`
+(🔥) but I am slightly leaning away from this given Protobuf is already used so
+heavily in our codebase. If we do use a different serialization format we'll
+need to be careful about how we encode numeric data, currently our JSON `Datum`
+uses `ProtoNumeric` which has very high precision.
+
+I am leaning away from this approach because we already use protobuf internally
+so it's well understood, and there are a few tricks we can use to improve
+deserialization to greatly improve our performance, e.g. zero-copy strings,
+lazy deserialization, and skipping fields we don't care about.
+
+[1] https://gist.github.com/ParkMyCar/594f647a1bc5a146bb54ca46e6e95680
+
+</td>
+
+<tr>
+<td> UUID </td>
+<td>
+
+```rust
+extern crate uuid;
+
+uuid::Uuid([u8; 16])
+```
+
+</td>
+<td>
+
+`FixedSizeBinary[16]`
+
+</td>
+<td>
+
+Encode the bytes from the `Uuid` directly into a fixed size buffer.
+
+</td>
+
+<tr>
+<td> Array </td>
+<td>
+
+```rust
+// DatumList<'a>.
+Vec<T>
+```
+
+</td>
+<td>
+
+`VariableSizeList<T>`
```

Comment:
Arrays are even a bit trickier than this. For each array value, you also need to encode the lower bounds, and the length of each dimension. And worth specifying whether you store the values in row-major or column-major order (in memory does row-major and it makes sense to keep that here.)

Reply:
Ah ha! They are definitely trickier than I thought. Just updated the design, I think I captured all of the nuance, I'm essentially proposing that we encode them like we do in proto today.